src/operator/contrib/transformer.cc

```diff
@@ -122,6 +122,531 @@ static bool InterleavedMatMulEncDecValAttShape(const NodeAttrs& attrs,
+                      head_dim,
+                      scale,
+                      keys_values,
+                      lead_dim_kv,
+                      batch_stride_kv,
+                      queries,
+                      lead_dim_q,
+                      batch_stride_q,
+                      beta,
+                      output,
+                      kv_seq_len,
+                      kv_seq_len * q_seq_len,
+                      attn_batches);
+}
+
+void BackwardInterleavedMatMulEncDecQKCPU(const nnvm::NodeAttrs& attrs,
+                                          const OpContext &ctx,
+                                          const std::vector<TBlob> &inputs,
+                                          const std::vector<OpReqType> &req,
+                                          const std::vector<TBlob> &outputs) {
+  const auto& params = nnvm::get<InterleavedMatMulParam>(attrs.parsed);
+  if (req[0] == kNullOp)
+    return;
+
+  CHECK_EQ(inputs[0].type_flag_, mshadow::kFloat32)
+    << "Only FP32 is supported on CPU at the moment";
+
+  mshadow::Stream<cpu>* s = ctx.get_stream<cpu>();
+  const float* output_grads = inputs[0].FlatTo2D<cpu, float>(s).dptr_;
+  const float* queries       = inputs[1].FlatTo2D<cpu, float>(s).dptr_;
+  const float* keys_values   = inputs[2].FlatTo2D<cpu, float>(s).dptr_;
+  float* queries_grads       = outputs[0].FlatTo2D<cpu, float>(s).dptr_;
+  float* keys_values_grads   = outputs[1].FlatTo2D<cpu, float>(s).dptr_;
+  const index_t q_seq_len         = inputs[1].shape_[0];
+  const index_t sequences         = inputs[1].shape_[1];
+  const index_t output_lin_q_dim  = inputs[1].shape_[2];
+  const index_t kv_seq_len        = inputs[2].shape_[0];
+  const index_t embed_dim         = output_lin_q_dim;
+  const index_t head_dim          = embed_dim / params.heads;
+  const index_t attn_batches      = params.heads * sequences;
+  const index_t lead_dim_q        = attn_batches * head_dim;
+  const index_t lead_dim_kv       = attn_batches * 2 * head_dim;
+  const index_t batch_stride_q    = head_dim;
+  const index_t batch_stride_kv   = head_dim * 2;
+  const float scale               = 1.f / sqrt(static_cast<float>(head_dim));
+
+  if (req[0] != kNullOp) {
+    const float beta = req[0] == kAddTo ? 1.f : 0.f;
+    strided_batch_sgemm(false,
+                        false,
+                        head_dim,
+                        q_seq_len,
+                        kv_seq_len,
+                        scale,
+                        keys_values,
+                        lead_dim_kv,
+                        batch_stride_kv,
+                        output_grads,
+                        kv_seq_len,
+                        kv_seq_len * q_seq_len,
+                        beta,
+                        queries_grads,
+                        lead_dim_q,
+                        batch_stride_q,
+                        attn_batches);
+  }
+  if (req[1] != kNullOp) {
+    if (req[1] == kWriteTo) {
```

Comment:
Shall we also check req[1] == kWriteInplace?

Reply:
Shall we also document why memset is required here

Reply:
> Shall we also check req[1] == kWriteInplace?

FInplaceOption is not registered in this backward operator. Do we still need check kWriteInplace here?

Reply:
> Shall we also document why memset is required here

Reference to the implementation on gpu path: https://github.com/apache/incubator-mxnet/blob/master/src/operator/contrib/transformer.cu#L160.
I thought it relates to the discussion about the bwd_ignore_zero_init flag and gradient accumulation in the original PR. @Caenorst any comments here?

Reply:
@Caenorst shall we change it to `req[1] == kWriteTo || req[1] == kWriteInplace ` for GPU code path, too?